src/bisimulation.rs

```diff
@@ -0,0 +1,373 @@
+            let splitter_block = match queue.pop() {
+                Some(coarse_block) => coarse_block,
+                None => break,
+            };
+            let mut fine_blocks_in_splitter_block = splitter_block
+                .fine_blocks_that_are_subsets_of_self
+                .borrow()
+                .clone();
+
+            let smaller_component_index = fine_blocks_in_splitter_block
+                .iter()
+                .enumerate()
+                .min_by(|(_, x), (_, y)| {
+                    (***x)
+                        .borrow()
+                        .values
+                        .len()
+                        .cmp(&(***y).borrow().values.len())
+                })
+                .map(|(index, _)| index)?;
+
+            let smaller_component = fine_blocks_in_splitter_block.remove(smaller_component_index);
+
+            let simple_splitter_block_values: Block = splitter_block
+                .values
+                .clone()
+                .iter()
+                .filter(|x| !(*smaller_component).borrow().values.contains(x))
+                .copied()
+                .collect();
+
+            let simple_splitter_block = CoarseBlock {
+                values: simple_splitter_block_values,
+                fine_blocks_that_are_subsets_of_self: RefCell::new(fine_blocks_in_splitter_block),
+            };
+            let simple_splitter_block_pointer = Rc::new(simple_splitter_block);
+
+            if simple_splitter_block_pointer
+                .fine_blocks_that_are_subsets_of_self
+                .borrow()
+                .len()
+                > 1
+            {
+                queue.push(Rc::clone(&simple_splitter_block_pointer));
+            }
+
+            (smaller_component, simple_splitter_block_pointer)
+        };
+        simple_splitter_block
+            .fine_blocks_that_are_subsets_of_self
+            .borrow()
+            .iter()
+            .for_each(|x| {
+                (*x).borrow_mut().coarse_block_that_supersets_self =
+                    Rc::clone(&simple_splitter_block);
+            });
+
+        let mut counterimage = build_counterimage(graph, smaller_component);
+
+        let counterimage_group = group_by_counterimage(counterimage.clone(), &node_to_block_vec);
+        let ((new_fine_blocks, removeable_fine_blocks), coarse_block_that_are_now_compound) =
+            split_blocks_with_grouped_counterimage(counterimage_group, &mut node_to_block_vec);
+
+        all_fine_blocks.extend(new_fine_blocks);
+        all_fine_blocks.retain(|x| !removeable_fine_blocks.iter().any(|y| Rc::ptr_eq(x, y)));
+        queue.extend(coarse_block_that_are_now_compound);
+
+        {
+            let counterimage_splitter_complement =
+                build_counterimage(graph, (*simple_splitter_block).clone());
+
+            counterimage_splitter_complement.keys().for_each(|node| {
+                counterimage.remove(node);
+            });
+        }
```

Comment:
Any particular statement this is in its own block and not with the other statements outside the bracket?

Reply:
I did this to contextually separate the $E^{-1}(B)-E^{-1}(S-B)$ operation. Should i add a comment or remove it?